### PR TITLE
Github Actions Added

### PR DESCRIPTION
**Pull request by AarC10**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,24 +1,17 @@
-##############################################################################
-# Convert CMake flags to compiler flags
 ###############################################################################
-if(EVT_CORE_LOG_ENABLE)
-    add_compile_definitions(EVT_CORE_LOG_ENABLE)
-endif()
-
-###############################################################################
-# Top level CMakeList for building the EVT HUDL source code
+# Top level CMakeList for building the EVT BMS source code
```
Comment (mjmagee991) [inline CMakeLists.txt]: This is unfortunately not the BMS.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,24 +1,17 @@
-##############################################################################
-# Convert CMake flags to compiler flags
 ###############################################################################
-if(EVT_CORE_LOG_ENABLE)
-    add_compile_definitions(EVT_CORE_LOG_ENABLE)
```
Comment (mjmagee991) [inline CMakeLists.txt:5]: Why did you delete this? This will prevent the logger from being enabled.

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,24 +1,17 @@
...
 
-# Link to the EVT-core library
-add_subdirectory(libs/EVT-core/)
-
 include(CMakeDependentOption)
 include(${EVT_CORE_DIR}/cmake/evt-core_compiler.cmake)
 include(${EVT_CORE_DIR}/cmake/evt-core_install.cmake)
 
+# Link to the EVT-core library
+add_subdirectory(libs/EVT-core/)
```
Comment (mjmagee991) [inline CMakeLists.txt]: Can you explain the problem this change solves? Should the project template be changed to match this?

File: include/HUDL/HUDL.hpp [hunk 4]
```diff
@@ -1,32 +1,38 @@
...
 class HUDL {
-public: 
-    
-    // default constructor for HUDL class
-    HUDL();
-  
-    // writes data to the LCD to show on the screen 
-    void data_write(unsigned char d);
+public:
+  IO::GPIO &reg_select; // PA_3
```
Comment (mjmagee991) [inline include/HUDL/HUDL.hpp]: Why are these members public?

File: include/HUDL/HUDL.hpp [hunk 5]
```diff
@@ -1,32 +1,38 @@
...
+  IO::GPIO &reset;      // PB_3
+  IO::GPIO &cs;         // PB_12
+  IO::SPI &spi;
 
-    // writes commands to the LCD to conrol the ST7565
-    void comm_write(unsigned char d);
+  // default constructor for HUDL class
+  HUDL(IO::GPIO &reg_select, IO::GPIO &reset, IO::GPIO &cs, IO::SPI &spi);
```
Comment (mjmagee991) [inline include/HUDL/HUDL.hpp]: This & spacing doesn't look like it matches our formatting standards. This might indicate a problem with the GitHub Action and/or the .clang-format file. See what you can do to fix this.

File: include/HUDL/HUDL.hpp [hunk 6]
```diff
@@ -1,32 +1,38 @@
...
 
-    // drives a single pixel on the LCD
-    void drive_pixel(unsigned char page, unsigned char col_up, unsigned char col_low, unsigned char data )
+  /**
+   * Writes data to the LCD to show on the screen
+   * @param data
```
Comment (mjmagee991) [inline include/HUDL/HUDL.hpp]: Can you update these docstrings to fit our normal standard?

File: include/HUDL/HUDL.hpp [hunk 7]
```diff
@@ -1,32 +1,38 @@
...
+  void clear_lcd(unsigned char *lcd_string);
 
+  // calls a set of commands to initialize LCD
+  void init_LCD();
+};
+} // namespace HUDL
 
-}
\ No newline at end of file
+#endif
```
Comment (mjmagee991) [inline include/HUDL/HUDL.hpp]: No newline :(

File: include/HUDL/dev/LED.hpp [hunk 8]
```diff
@@ -7,41 +7,38 @@ namespace EVT::core::DEV {
 
 class LED {
```
Comment (mjmagee991) [inline include/HUDL/dev/LED.hpp]: I'm pretty sure we have an LED class in EVT-core. Do you know why this is here?

File: include/HUDL/dev/LED.hpp [hunk 9]
```diff
@@ -7,41 +7,38 @@ namespace EVT::core::DEV {
...
-    /**
-     * Set the current state of the LED.
-     *
-     * @param state[in] The state to set the LED to.
-     */
-    void setState(EVT::core::IO::GPIO::State state);
+  /**
+   * Represents if the LED is active high or active low.
+   */
+  enum class ActiveState { HIGH = 0u, LOW = 1u };
```
Comment (mjmagee991) [inline include/HUDL/dev/LED.hpp]: This also doesn't look formatted correctly.

File: src/HUDL/HUDL.cpp [hunk 10]
```diff
@@ -3,176 +3,119 @@
...
  */
-#include <stdint.h>
 
-#include <EVT/io/GPIO.hpp>
-#include <EVT/io/I2C.hpp>
-#include <EVT/io/manager.hpp>
-#include <EVT/io/pin.hpp>
-#include <EVT/io/UART.hpp>
-#include <EVT/utils/time.hpp>
+// clang-format off
```
Comment (mjmagee991) [inline src/HUDL/HUDL.cpp]: Can you add a comment explaining why this is necessary?

Comment (mjmagee991) [reply src/HUDL/HUDL.cpp]: This might actually not be necessary after you fix the formatting file.

File: src/HUDL/HUDL.cpp [hunk 11]
```diff
@@ -3,176 +3,119 @@
...
-void HUDL::comm_write(unsigned char d) 
-{
-    d = (uint8_t)d;
-    CS.writePin(EVT::core::IO::GPIO::State::LOW);
-    reg_select.writePin(EVT::core::IO::GPIO::State::LOW);
-    spi.startTransmission(0);
-    spi.write(&d, 1);
-    spi.endTransmission(0);
-    CS.writePin(EVT::core::IO::GPIO::State::HIGH);
+/**
```
Comment (mjmagee991) [inline src/HUDL/HUDL.cpp]: These docstrings should be in hpp but not cpp.

File: src/HUDL/HUDL.cpp [hunk 12]
```diff
@@ -3,176 +3,119 @@
...
-        //ClearLCD();
-
-        board.drive_pixel(1, 1, 1, 255);
-        time::wait(10000); 
-
-        //echos command back
-        uart.printf("\n\recho: %s\n\r", buf);
-    }
-}
+} // namespace HUDL
```
Comment (mjmagee991) [inline src/HUDL/HUDL.cpp]: No newline :(

File: targets/HUDL/main.cpp [hunk 13]
```diff
@@ -6,185 +6,64 @@
...
-    comm_write(0xA2);   // LCD bias set
-    comm_write(0x2F);   // Power Control set
-    comm_write(0x26);   // Resistor Ratio Set 
-    comm_write(0x81);   // Electronic Volume Command (set contrast) Double Btye: 1 of 2
-    comm_write(0x11);   // Electronic Volume value (contrast value) Double Byte: 2 of 2
-    comm_write(0xAF);   // Display ON
-}
-
-
+const uint32_t SPI_SPEED = SPI_SPEED_4MHZ; // 4MHz
```
Comment (mjmagee991) [inline targets/HUDL/main.cpp]: Why const instead of constexpr?

File: targets/HUDL/main.cpp [hunk 14]
```diff
@@ -6,185 +6,64 @@
...
 int main() {
-    
-    //HUDL board = new HUDL();
-    HUDL *board = new HUDL();
-     //register select set
-    board->reg_select = IO::getGPIO<IO::Pin::PA_3>(EVT::core::IO::GPIO::Direction::OUTPUT);
-    board->reg_select.writePin(EVT::core::IO::GPIO::State::LOW);
+  IO::GPIO *devices[deviceCount];
+  auto &reg_select =
+      IO::getGPIO<IO::Pin::PA_3>(EVT::core::IO::GPIO::Direction::OUTPUT);
```
Comment (mjmagee991) [inline targets/HUDL/main.cpp]:
```suggestion
      IO::getGPIO<IO::Pin::PA_3>(IO::GPIO::Direction::OUTPUT);
```

Comment (mjmagee991) [reply targets/HUDL/main.cpp]: Just so it can fit in 1 line

File: targets/HUDL/main.cpp [hunk 15]
```diff
@@ -6,185 +6,64 @@
...
 
-        board->drive_pixel(1, 1, 1, 255);
-        time::wait(10000); 
+    board.drive_pixel(1, 1, 1, 255);
+    time::wait(10000);
 
-        //echos command back
-        uart.printf("\n\recho: %s\n\r", buf);
-    }
+    // echos command back
```
Comment (mjmagee991) [inline targets/HUDL/main.cpp]:
```suggestion
    // echoes command back
```
_**e**_

File: targets/HUDL/write_to_display.cpp [hunk 16]
```diff
@@ -0,0 +1,71 @@
+//
```
Comment (mjmagee991) [inline targets/HUDL/write_to_display.cpp]: Replace this with a standard comment block explaining the purpose of this file.

File: targets/HUDL/write_to_display.cpp [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+      IO::getGPIO<IO::Pin::PB_3>(EVT::core::IO::GPIO::Direction::OUTPUT);
+
+  // reset the board
+  reset.writePin(EVT::core::IO::GPIO::State::LOW);
+  time::wait(100);
+  reset.writePin(EVT::core::IO::GPIO::State::HIGH);
+  time::wait(100);
+
+  // cs
+  auto &cs =
```
Comment (mjmagee991) [inline targets/HUDL/write_to_display.cpp]: What is cs?

File: targets/HUDL/write_to_display.cpp [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+#include <HUDL/HUDL.hpp>
+
+namespace IO = EVT::core::IO;
+namespace time = EVT::core::time;
+
+constexpr uint32_t SPI_SPEED = SPI_SPEED_4MHZ; // 4MHz
+constexpr uint8_t deviceCount = 1;
+
+int main() {
+  IO::GPIO *devices[deviceCount];
```
Comment (mjmagee991) [inline targets/HUDL/write_to_display.cpp]: This should propbably references instead of pointers.

**Review by mjmagee991 — CHANGES_REQUESTED** [review]
That was a lot of code. It's still a bit messy, but it's been formalized much more than it was. Take a look at my comments, and hopefully we'll be able to get it standardized quickly.

File: CMakeLists.txt (hunk 3 above)
Comment (AarC10) [reply CMakeLists.txt]: Must've been a mistake. It has been placed back where it was originally.

File: include/HUDL/dev/LED.hpp (hunk 8 above)
Comment (AarC10) [reply include/HUDL/dev/LED.hpp]: Unknown why it was there in the first place. Looks like it was not reference in CMake, so it is now removed.

File: src/HUDL/HUDL.cpp (hunk 10 above)
Comment (AarC10) [reply src/HUDL/HUDL.cpp]: This is just a quick band-aid fix to GitHub reformatting and causing compile errors. The issue will be researched eventually and fixed.

File: targets/HUDL/main.cpp (hunk 13 above)
Comment (AarC10) [reply targets/HUDL/main.cpp]: SPI_SPEED_4MHZ is just a macro defined integer meaning the value is known at compile time. Therefore SPI_SPEED would still be a compile time constant.

File: targets/HUDL/write_to_display.cpp (hunk 18 above)
Comment (AarC10) [reply targets/HUDL/write_to_display.cpp]: Can you explain more? Why should this be a reference if we are initializing a buffer to store a device.

File: include/HUDL/HUDL.hpp (hunk 7 above)
Comment (S-TH3TA-13) [reply include/HUDL/HUDL.hpp]: sadge

File: targets/HUDL/write_to_display.cpp [hunk 19]
```diff
@@ -12,11 +12,11 @@
 namespace IO = EVT::core::IO;
 namespace time = EVT::core::time;
 
-constexpr uint32_t SPI_SPEED = SPI_SPEED_4MHZ;// 4MHz
-constexpr uint8_t deviceCount = 1;
+const uint32_t SPI_SPEED = SPI_SPEED_4MHZ;// 4MHz
+const uint8_t deviceCount = 1;
```
Comment (S-TH3TA-13) [inline targets/HUDL/write_to_display.cpp]: change const unit8_t deviceCount =1 to captitals since it is a constant. ie. const unit8_t DEVICECOUNT = 1 
Need to then replace name in rest of main()
(๑ↀᆺↀ๑)

**Comment by AarC10** [comment]
L @DeLozM8

**Review by S-TH3TA-13 — CHANGES_REQUESTED** [review]

File: targets/HUDL/write_to_display.cpp (hunk 19 above)
Comment (AarC10) [reply targets/HUDL/write_to_display.cpp]: image

Comment (S-TH3TA-13) [reply targets/HUDL/write_to_display.cpp]: DEVICE_COUNT
ლ(●ↀωↀ●)ლ

**Review by S-TH3TA-13 — COMMENTED** [review]
pog ty that was my bad i guess

**Review by ActuallyTaylor — APPROVED** [review]
Looks all good 👍

File: include/HUDL/HUDL.hpp [hunk 20]
```diff
@@ -1,32 +1,59 @@
...
-    // drives a single pixel on the LCD
-    void drive_pixel(unsigned char page, unsigned char col_up, unsigned char col_low, unsigned char data )
-
-    // clears the LCD 
-    void ClearLCD(unsigned char *lcd_string);
-
-    // calls a set of commands to initialize LCD
+public:
+    /**
+    * Default Constructor for the HUDL class
```
Comment (mjmagee991) [inline include/HUDL/HUDL.hpp:16]: There should be a blank line between the description and parameters

File: include/HUDL/HUDL.hpp [hunk 21]
```diff
@@ -1,32 +1,59 @@
...
+    * @param reg_select is the register select pin
+    * @param reset is the reset pin
```
Comment (mjmagee991) [inline include/HUDL/HUDL.hpp]: All the parameters should be marked with `@param[in]`

File: src/HUDL/HUDL.cpp (hunk 10 above)
Comment (mjmagee991) [reply src/HUDL/HUDL.cpp]: Can you add a comment that says that? It would probably also make sense to make a Trello card for that to ensure that doesn't get forgotten down the line

**Review by mjmagee991 — CHANGES_REQUESTED** [review]
Just had a couple comments. I'll do a deeper check after you pull the EVT-core changes

File: targets/HUDL/main.cpp (hunk 13 above)
Comment (mjmagee991) [reply targets/HUDL/main.cpp]: That makes sense, but `constexpr` is the keyword used for compile-time constants, not `const`. I read some info here: https://stackoverflow.com/questions/13346879/const-vs-constexpr-on-variables

File: targets/HUDL/write_to_display.cpp (hunk 18 above)
Comment (mjmagee991) [reply targets/HUDL/write_to_display.cpp]: Pointers are more dangerous than references. If you can instantiate as a reference instead of a pointer, you should. https://stackoverflow.com/questions/7058339/when-to-use-references-vs-pointers

**Review by S-TH3TA-13 — APPROVED** [review]

File: include/HUDL/HUDL.hpp [hunk 22]
```diff
@@ -1,6 +1,11 @@
 #ifndef _HUDL_
 #define _HUDL_
 
+#include "EVT/dev/LCD.hpp"
```
Comment (mjmagee991) [inline include/HUDL/HUDL.hpp]: Change to `<>` includes

File: include/HUDL/HUDL.hpp [hunk 23]
```diff
@@ -18,51 +23,49 @@ class HUDL {
...
-    void init_LCD();
+    void initLCD();
 
 private:
-    // TODO: Remove later. Commented out since it is used for LCD now
-//    IO::GPIO& reg_select;// PA_3
-//    IO::GPIO& reset;     // PB_3
-//    IO::GPIO& cs;        // PB_12
-//    IO::SPI& spi;
+    /**
```
Comment (mjmagee991) [inline include/HUDL/HUDL.hpp]: Can you delete these?

File: src/HUDL/HUDL.cpp [hunk 24]
```diff
@@ -12,31 +12,30 @@
...
+    lcd.clearLCD(bitmap);
 }
 
-void HUDL::init_LCD() {
-
+void HUDL::initLCD() {
+    lcd.initLCD();
 }
 
 }// namespace HUDL
```
Comment (mjmagee991) [inline src/HUDL/HUDL.cpp]: Add an ending newline

**Review by mjmagee991 — CHANGES_REQUESTED** [review]
Just a couple minor problems. Overall, great PR

File: include/HUDL/HUDL.hpp (hunk 22 above)
Comment (AarC10) [reply include/HUDL/HUDL.hpp]: Kind of weird to use <> for non standard library headers but sure?

**Review by mjmagee991 — APPROVED** [review]
Great work. I really like how this turned out
